toxcore/tox.c

```diff
@@ -328,6 +328,51 @@ bool tox_version_is_compatible(uint32_t major, uint32_t minor, uint32_t patch)
     return TOX_VERSION_IS_API_COMPATIBLE(major, minor, patch);
 }
 
+static State_Load_Status state_load_callback(void *outer, const uint8_t *data, uint32_t length, uint16_t type)
+{
+    Tox *tox = (Tox *)outer;
+    State_Load_Status status = STATE_LOAD_STATUS_CONTINUE;
+
+    if (messenger_load_state_section(tox->m, data, length, type, &status)
+            || conferences_load_state_section(tox->m->conferences_object, data, length, type, &status)) {
+        return status;
+    }
+
+    if (type == TOP_STATE_TYPE_END) {
+        if (length != 0) {
+            return STATE_LOAD_STATUS_ERROR;
+        }
+
+        return STATE_LOAD_STATUS_END;
+    }
+
+    LOGGER_ERROR(tox->m->log, "Load state: contains unrecognized part (len %u, type %u)\n",
+                 length, type);
+
+    return STATE_LOAD_STATUS_CONTINUE;
+}
+
+/* Load tox from data of size length. */
+static int tox_load(Tox *tox, const uint8_t *data, uint32_t length)
+{
+    uint32_t data32[2];
+    uint32_t cookie_len = sizeof(data32);
+
+    if (length < cookie_len) {
+        return -1;
+    }
+
+    memcpy(data32, data, sizeof(uint32_t));
+    lendian_bytes_to_host32(data32 + 1, data + sizeof(uint32_t));
+
+    if (!data32[0] && (data32[1] == STATE_COOKIE_GLOBAL)) {
```

Comment:
Replace `!num` with `num == 0` when `num` is an integer, not a bool. Also, would it make sense to return -1 on the inverted condition, considering it an error return, and returning the success/failure of state_load at the end of the function?